### PR TITLE
INT: Add IfLetToMatchIntention

### PR DESCRIPTION
**Pull request by Shiroy**
PR for issue #962 

This intention convert
```rust
if let A(value) = x {
    //Stuff for A
} else if let B(value) = x {
    //Stuff for B
} else {
    //Stuff for something else
}
```
to
```rust
match x {
    A(value) => {
        //Stuff for A
    },
    B(value) => {
        //Stuff for B
    },
    _ => {
        //Stuff for something else
    }
}
```

### CODE REVIEW

**Comment by Shiroy** [comment]
Is there a way I can write unit tests for that intention

**Comment by matklad** [comment]
>Is there a way I can write unit tests for that intention?

Yes, they are in /home/matklad/projects/intellij-rust/src/test/kotlin/org/rust/ide/intentions

`doAvailableTest` takes two string literals with Rust code and a caret marker `/*caret*/` and checks that after applying the intention to the first string, we get the second one.

`doUnavailableTest` checks that the intention is not available. 


I haven't reviewed the code in detail, because it is WIP, but it looks good on the first blush! Could you run a **reformat file** action though? We place a space between `if` and `(` in condition :)

File: src/main/kotlin/org/rust/ide/intentions/IfLetToMatchIntention.kt [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+        //Here we have extracted the upper most if statement node
+
+        return extractIfLetStatementIfAny(ifStatement)
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val (ifStmt, target, matchArms, elseBody) = ctx
+
+        var generatedCode = "match ${target.text} {"
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/intentions/IfLetToMatchIntention.kt]: You can use `buildString` for more efficient and idiomatic code:

```
        buildString { 
            append("foo")
            append("bar")
        }
```

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/build-string.html

**Comment by Shiroy** [comment]
I'm done :)

**Comment by matklad** [comment]
Super!

**Comment by matklad** [comment]
(this will be in nightly in 24 hours, and in alpha the next monday).

**Comment by alygin** [comment]
I believe, there must be an intention description in `src/main/resources/intentionDescriptions`. Without it, I got the following error:

```
Assertion failed: Intention Description Dir URL is null: Convert if let statement to match; IfLetToMatchIntention, PluginClassLoader[org.rust.lang, 0.1.0.SNAPSHOT]
java.lang.Throwable
	at com.intellij.openapi.diagnostic.Logger.assertTrue(Logger.java:168)
	at com.intellij.codeInsight.intention.impl.config.IntentionActionMetaData.getDirURL(IntentionActionMetaData.java:109)
	at com.intellij.codeInsight.intention.impl.config.BeforeAfterActionMetaData.getDescription(BeforeAfterActionMetaData.java:155)
```

**Comment by matklad** [comment]
Yep, I need to add a test for this.

**Comment by matklad** [comment]
https://github.com/intellij-rust/intellij-rust/commit/f3ca2f638e501558b9e35d5cbdafbaecae32c733

**Pull request merged**